File: 2021/matlab/aoc2.m
txt = read_txt(fullfile(get_input_root(),'day_2.txt'));
lines = strsplit(txt,'\n');

parse_moves = @(lines,key) (str2double(extractAfter(lines(contains(lines,key)),' ')));
f = parse_moves(lines,"forward"); 
d = parse_moves(lines,"down"); 
u = parse_moves(lines,"up"); 

d_m = sum(d) - sum(u);
f_m = sum(f);

%part 1
p1 = d_m*f_m;
disp(p1)
assert(p1 == 1459206)

aim = 0;
h = 0;
d = 0;
for i1 = 1:numel(lines)
    x = str2double(extractAfter(lines{i1},' '));
    if contains(lines{i1},"down")
        aim = aim + x;
    end
    
    if contains(lines{i1},"up")
        aim = aim - x;
    end
    
    if contains(lines{i1},"forward")
        h = h + x;
        d = d  + aim*x;
    end
end

%part 2
p2 = h*d;
disp(int32(p2))
assert(p2 == 1320534480)

File: 2021/python/aoc2.py
input_path = '../inputs/day_2.txt'
lines = open(input_path).read().split('\n')

move = [float(x.split(' ')[-1]) for x in lines]
direction = [x.split(' ')[0] for x in lines]
move_pairs = list(zip(direction,move));

move = [-m if d == 'up' else m for d,m in move_pairs]
direction = ['d' if d in {'up','down'} else 'h' for d in direction]

move_pairs = list(zip(direction,move));

dh = dict(zip(['h','d'],[0,0]))

for k,v in move_pairs: dh[k] += v

p1 = dh['d']*dh['h']
print(p1)

#part 2
dha= dict(zip(['h','d','a'],[0,0,0]))

for k,v in move_pairs: 
    if k == 'd':
        dha['a'] += v
    elif k == 'h':
        dha['h'] += v
        dha['d'] += v*dha['a']
        
p2 = dha['d']*dha['h']
print(p2)